### PR TITLE
Add generator module for protoc plugin development

### PR DESCRIPTION
**Pull request by amery**
# Add generator module for protoc plugin development

## Summary

Introduces a new `generator` submodule providing essential utilities for
developing protocol buffer compiler (protoc) plugins. This module offers
type-safe descriptor handling, comprehensive test utilities, and follows
strict TDD principles with 100% test coverage.

## Motivation

The protomcp.org ecosystem requires robust utilities for working with
protocol buffer descriptors when building code generators. This module
provides the foundational layer that other modules (particularly the
planned `options` module) will depend upon for descriptor traversal,
type checking, and manipulation.

### Core Functionality

The module provides three main categories of utilities:

#### 1. Descriptor Type Checking (descriptor.go)

- Type casting functions (`AsFoo`) for safe conversion to specific
  descriptor types with essential field validation.
- Boolean checking functions (`IsFoo`) for type verification.
- Support for all major descriptor types: File, Message, Field, Enum,
  Service, and Method.
- Special `AsMessageWithName`/`IsMessageWithName` for name-based matching.
- Field characteristic helpers for cardinality (repeated, map, oneof)
  and classification (scalar, message, group, enum).
- Map field detection with both heuristic (`AsMapField`) and definitive
  (`AsMapFieldWithMessage`) checking.
- Proto3Optional field semantics documentation.
- Validation of essential fields (Name, Type, etc.) in all AsXType functions.
- Separate handling for Group types (deprecated in proto2, not in proto3).

#### 2. Test Utilities (testutils.go, types.go)

- Complete field constructors for creating fully-specified test
  descriptors.
- Minimal field constructors for testing specific properties in
  isolation.
- Helper functions for creating messages, enums, services, and files.
- Type constants for all protobuf field types and labels.

#### 3. Comprehensive Testing

Test files: descriptor_test.go, testutils_test.go, types_test.go

- Table-driven tests following darvaza.org/core patterns.
- Factory pattern for test case generation.
- Edge case coverage for robust validation.
- 100% test coverage achieved and maintained.

### Documentation

- Comprehensive README with usage examples and API reference.
- Package documentation in doc.go.
- Clear organisation of functions into logical categories.
- Tables for better readability of function signatures.

### Implementation Progression

The module was developed incrementally following TDD principles across
three commits:

1. **Placeholder**: Established initial module structure with planned
   architecture.
2. **Test utilities**: Created comprehensive test utilities for descriptor
   creation, including factory functions and type constants.
3. **Core functionality**: Implemented type checking and casting functions
   with 100% test coverage.

## Testing

All functionality has been thoroughly tested:

- **Coverage**: 100% test coverage across all files.
- **Patterns**: Consistent use of table-driven tests.
- **Organisation**: Factory functions for test case generation.
- **Edge cases**: Comprehensive coverage including nil inputs and invalid
  types.

## Impact

This module provides the foundation for:

- Building robust protoc plugins.
- Implementing the options override system.
- Creating type-safe descriptor traversal utilities.
- Standardising test creation across the protomcp.org ecosystem.

## Future Work

The README documents planned features that will be added incrementally:

- Path construction and naming utilities.
- Descriptor traversal functions (Find, ForEach patterns).
- Visitor pattern for walking descriptor trees.
- Context management for build environments.
- Code generation output helpers.

## Breaking Changes

None. This is a new module with no existing consumers.

## Dependencies

Minimal external dependencies:

- `google.golang.org/protobuf` for descriptor types.
- `darvaza.org/core` for test utilities only (not runtime).

## Review Notes

- All functions follow consistent `AsFoo`/`IsFoo` naming patterns.
- Nil safety is guaranteed across all public APIs.
- British English spelling used throughout documentation.
- Line lengths kept under 80 characters for readability.
- Comprehensive documentation at package, function, and example levels.

___

### **PR Type**

Enhancement

___

### **Description**

- Add comprehensive descriptor type checking utilities with AsFoo/IsFoo
  pattern.

- Implement field classification (scalar, message, enum) and cardinality
  helpers (repeated, map, oneof).

- Create extensive test utilities with factory functions and type constants.

- Achieve 100% test coverage with table-driven tests following darvaza.org
  patterns.

- Graduate module from placeholder to active development.

___

### Diagram Walkthrough

```mermaid
flowchart LR
  A["generator module"] --> B["descriptor.go"]
  A --> C["testutils.go"]
  A --> D["types.go"]
  A --> E["doc.go"]
  B --> F["Type checking<br/>AsFoo/IsFoo"]
  B --> G["Field classification<br/>Scalar/Message/Enum"]
  B --> H["Field cardinality<br/>Repeated/Map/OneOf"]
  C --> I["Test constructors<br/>NewField variants"]
  C --> J["Minimal constructors<br/>NewFieldWithType"]
  D --> K["Type constants<br/>TypeString, TypeInt32"]
  D --> L["Label constants<br/>LabelOptional, LabelRepeated"]
```

<details> <summary><h3> File Walkthrough</h3></summary>

<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>descriptor.go</strong><dd><code>Implement descriptor type checking and field classification</code></dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-e47c5102ba35780462b0d73fcf9d5a61dd12a8c2">+356/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>testutils.go</strong><dd><code>Create test utilities for descriptor construction</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-dc72e9c68f425e94739e504e0915c0cfe3112466">+235/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>types.go</strong><dd><code>Define type and label constants for protobuf fields</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-385f528dd9a7bf836d89c17f8aee54c91ea11797">+41/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Tests</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td><strong>descriptor_test.go</strong><dd><code>Add comprehensive tests with factory pattern</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-604787013108692bfad9d9ffabc04a49b74473f4d3124d95769e0cd0962dd942">+425/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>testutils_test.go</strong><dd><code>Test all utility functions and edge cases</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-f29d023679c12f67913a58ff1800553c7c4028f3cdfb54baaa4410f953beaecc">+417/-0</a>&nbsp; </td>

</tr>

<tr>
  <td><strong>types_test.go</strong><dd><code>Test type and label constants</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-types_test">+41/-0</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>4 files</summary><table>
<tr>
  <td><strong>doc.go</strong><dd><code>Add package documentation with API reference</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-5b575806537dd8e15943da90f327934c25f46b9d">+52/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>AGENT.md</strong><dd><code>Document generator submodule in development guide</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-c1e87407ce72f38b645f627c7fbda6e6999170c840cded6ce82c1bd2c10923e8">+9/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>README.md</strong><dd><code>Add generator utilities to main README</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>generator/README.md</strong><dd><code>Create comprehensive module documentation with examples</code>&nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-a715bdcf88cd771d7ae09bd01863bab314ff194ba28e984689e9040d87846db9">+270/-194</a>&nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Miscellaneous</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>LICENCE.txt</strong><dd><code>Add MIT licence for generator submodule</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-c0f8bd37207bd6cb420debc71fcfa003d8164d2827876ed6e0d233b947c71656">+19/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Configuration changes</strong></td><td><details><summary>2 files</summary><table>
<tr>
  <td><strong>go.mod</strong><dd><code>Initialise Go module with dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-b827f55ed9eb4afc993ba2d44b448b74dfb6177deb13508ab36515c4c4341938">+13/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td><strong>cspell.json</strong><dd><code>Add protobuf-related terms to spell checker</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-3094120fcb2e87cf154fce1053bc486a13eaa9d4200a24f858bc9c39ec2e0936">+6/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></details></td></tr><tr><td><strong>Dependencies</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td><strong>go.sum</strong><dd><code>Add dependency checksums for module</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></td>
  <td><a href="https://github.com/protomcp/common/pull/3/files#diff-874492ac35aad47ffa54891f67d2ff868c4470a4112641bfb961ce45ea6bffa2">+12/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></details></td></tr></tr></tbody></table>

</details>

___

## Summary by CodeRabbit

- **New Features**
  - Adds a Generator module with descriptor inspection, field classification (repeated, map, oneof, optional/required), type-checking helpers, type alias constants, and builder/test utilities for protoc plugin development.

- **Documentation**
  - Adds comprehensive Generator docs, package-level docs, generator README, and updates main README with a Go Report Card badge and “Generator Utilities” entry plus usage examples.

- **Tests**
  - Adds extensive unit tests covering descriptor predicates, builder utilities, and type/label aliases.

- **Chores**
  - Adds MIT license for Generator, initializes module metadata, and updates internal spelling dictionary.

### CODE REVIEW

File: generator/go.mod [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module protomcp.org/common/generator
+
+go 1.23.0
```
[the remark this answers is not in the document]
Comment (amery) [reply generator/go.mod:3]: invalid. go 1.23 requires X.Y.Z there

**Comment by amery** [comment]
# PR Update: Enhanced Descriptor Type Checking

Based on review feedback, enhanced descriptor utilities to be more robust and useful.

## Key Changes

### Essential Field Validation
- All `AsXType` functions now validate required fields (Name, Type, etc.)
- Descriptors without essential fields are considered invalid
- Functions return nil for invalid descriptors rather than zero-values

### New Functions
- Added `AsMapFieldWithMessage`/`IsMapFieldWithMessage` for definitive map field checking via `map_entry=true` option
- Added `AsGroupField`/`IsGroupField` to separate Group from Message types (Groups deprecated in proto2)
- Added `isPointerEqual` and `isPointerNonZero` generic helpers

### Improvements
- Proto3Optional field semantics documented
- Fixed `NewFieldWithLabel` to include default Type
- Updated all tests to verify field validation
- Fixed field alignment in test structs

### Documentation
- Added new functions to README.md and doc.go
- Fixed punctuation consistency
- Corrected AGENT.md submodule description

All tests pass with 100% coverage.

**Comment by amery** [comment]
/review
--pr_reviewer.inline_code_comments=true
